src/routes/blog/post/remix-3-whats-changing-and-why-it-matters/+page.markdoc

```diff
@@ -0,0 +1,144 @@
+
+```ts
+export const routes = route({
+  books: {
+    index: "/books",
+    show: "/books/:slug"
+  }
+});
+```
+
+From there, Remix can infer types and generate helpers like `routes.books.show.href({ slug: "gatsby" })`. This keeps links type-safe and prevents mistakes. It also lets the router generate correct URLs for any environment, whether the app runs on Node or Cloudflare Workers.
+
+### The new rendering model
+
+The biggest change is that Remix 3 no longer uses React. You still write JSX, but it's handled by a small custom runtime built for Remix itself. There's no virtual DOM. When something changes, you update state directly and call `this.update()` to refresh that part of the UI. Events use the browser's native event system instead of React's synthetic one.
+
+This makes the rendering model closer to plain JavaScript. Developers can see exactly what's happening without guessing at hidden framework behavior. The trade-off is that you need to be more deliberate about updates, there's no `useState` hook to trigger automatic re-renders. But it also means fewer performance surprises and much less overhead.
+
+To handle partial updates, Remix 3 introduces a feature called Frames. A Frame is a section of the page that can load and update independently. When a Frame refreshes, the server sends back a small HTML fragment, and Remix patches it into the DOM. This allows streaming updates without extra JavaScript or JSON APIs.
+
+For example, a comments section could live inside a Frame. When a user submits a new comment, the server renders updated HTML for that frame and sends it down. The rest of the page stays untouched. It feels like client-side interactivity, but it's still server-driven and progressively enhanced.
+
+### Data loading and actions
+
+The loader and action pattern that defined Remix 2 is still here. Each route can export handlers for different HTTP methods. A `GET` handler fetches data to render; a `POST` (or `PUT`, `DELETE`) handler performs a mutation.
+
+Here's a simple example:
+
+```ts
+export const handlers = {
+  async GET({ params }) {
+    const post = await getPost(params.slug);
+    if (!post) return render(<NotFound />, { status: 404 });
+    return render(<PostPage post={post} />);
+  },
+
+  async POST({ request, params }) {
+    const form = await request.formData();
+    const comment = form.get("comment");
+    await saveComment(params.slug, comment);
+    return redirect(routes.posts.show.href({ slug: params.slug }));
+  }
+};
+```
+
+In this setup, the `GET` handler runs on the server and renders HTML for the initial page. The `POST` handler processes a submitted `<form>`, updates data, and redirects. Remix then re-runs the necessary loaders to refresh the UI. If JavaScript is disabled, the browser still handles the POST and redirect normally, since everything uses standard HTML forms.
```

Comment:
This is very good point, but very late in article. Somewhere at top, I would mention It uses JavaScript, but on server side. Meaning on client side it utilizes existing browser capabilities, making it more accessible to parsers, but also to some devices and browsers with limited JavaScript support